lib/jxl/enc_ac_strategy.cc

```diff
@@ -659,6 +713,149 @@ void FindBest16X16(size_t bx, size_t by, size_t cx, size_t cy,
   }
 }
 
+// The following function tries to merge 8x8 transforms into
+// 32X16 and 16X32 DCTs fairly, by trying them and their combinations
+// with best 16x16 (including the best 16X16 subdivision because of
+// the bottom up merge approach) at the same time.
+//
+// TODO(jyrki):
+// This idea could be generalized to larger transforms.
+void FindBest32X32(size_t bx, size_t by, size_t cx, size_t cy,
+                   const ACSConfig& config,
+                   const float* JXL_RESTRICT cmap_factors,
+                   AcStrategyImage* JXL_RESTRICT ac_strategy,
+                   const float entropy_mul, const float entropy_mul_32X32,
+                   float* JXL_RESTRICT entropy_estimate, float* block,
+                   float* scratch_space, uint32_t* quantized) {
+  constexpr AcStrategy::Type acs_raw32X16 =
+      AcStrategy::Type::DCT32X16;  // y=32, x=16
+  constexpr AcStrategy::Type acs_raw16X32 =
+      AcStrategy::Type::DCT16X32;  // y=16, x=32
+  constexpr AcStrategy::Type acs_raw32X32 = AcStrategy::Type::DCT32X32;
+  const AcStrategy acs32X16 = AcStrategy::FromRawStrategy(acs_raw32X16);
+  const AcStrategy acs16X32 = AcStrategy::FromRawStrategy(acs_raw16X32);
+  const AcStrategy acs32X32 = AcStrategy::FromRawStrategy(acs_raw32X32);
+  AcStrategyRow row0 = ac_strategy->ConstRow(by + cy + 0);
+  AcStrategyRow row1 = ac_strategy->ConstRow(by + cy + 2);
+  {
+    bool has32X32 = row0[bx + cx + 0].RawStrategy() == acs_raw32X32;
+    if (has32X32) {
+      return;
+    }
+  }
+  if (MultiBlockTransformCrossesHorizontalBoundary(*ac_strategy, bx + cx,
+                                                   by + cy, bx + cx + 4) ||
+      MultiBlockTransformCrossesHorizontalBoundary(*ac_strategy, bx + cx,
+                                                   by + cy + 4, bx + cx + 4) ||
+      MultiBlockTransformCrossesVerticalBoundary(*ac_strategy, bx + cx, by + cy,
+                                                 by + cy + 4) ||
+      MultiBlockTransformCrossesVerticalBoundary(*ac_strategy, bx + cx + 4,
+                                                 by + cy, by + cy + 4)) {
+    return;  // not suitable for 32x32 analysis, some transforms leak out.
+  }
+  const bool allow_32X16 = !MultiBlockTransformCrossesVerticalBoundary(
+      *ac_strategy, bx + cx + 2, by + cy, by + cy + 4);
+  const bool allow_16X32 = !MultiBlockTransformCrossesHorizontalBoundary(
+      *ac_strategy, bx + cx, by + cy + 2, bx + cx + 4);
+  // Current entropies aggregated on 16x16 resolution.
+  float entropy[2][2] = {};
+  for (size_t dy = 0; dy < 4; ++dy) {
+    for (size_t dx = 0; dx < 4; ++dx) {
+      entropy[dy >> 1][dx >> 1] += entropy_estimate[(cy + dy) * 8 + (cx + dx)];
+    }
+  }
+  float try32X16_0 = std::numeric_limits<float>::max();
```

Comment:
This is a bad name for the variable

Reply:
renamed more descriptive (with left/right/top/bottom instead of 0/1)

Reply:
I was more complaining about "try"

Reply:
both fixed